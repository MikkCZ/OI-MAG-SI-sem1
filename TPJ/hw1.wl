(* ::Package:: *)

isRewritable[e_]:=(Or[Head[oneStepRewrite[e]]=!=oneStepRewrite,e==={}]);

notIntNotSym[s_Symbol]:=False;
notIntNotSym[n_Integer]:=False;
notIntNotSym[_]:=True;

(* Number *)
oneStepRewrite[derivative[n_Integer]]:={0};

(* Symbol - "x" *)
oneStepRewrite[derivative[x]]:={1};

(* Addition - "plus" *)
(* Your code here... *)
oneStepRewrite[plus[e_?PossibleZeroQ,f_]]:={f};
oneStepRewrite[plus[e_,f_?PossibleZeroQ]]:={e};
oneStepRewrite[plus[e_Integer,f_Integer]]:={Plus[e,f]};
oneStepRewrite[plus[e_,f_]] /; notIntNotSym[e]&&notIntNotSym[f] := {plus[oneStepRewrite[e][[1]],f],plus[e,oneStepRewrite[f][[1]]]};

(* Multiplication - "times" *)
(* Your code here... *)
oneStepRewrite[times[e_?PossibleZeroQ,f_]]:={0};
oneStepRewrite[times[e_,f_?PossibleZeroQ]]:={0};
oneStepRewrite[times[e_Integer,f_Integer]]:={Times[e,f]};
oneStepRewrite[times[x,f_]] /; notIntNotSym[f] := {times[x,oneStepRewrite[f][[1]]]};
oneStepRewrite[times[e_,x]] /; notIntNotSym[e] := {times[oneStepRewrite[e][[1]],x]};
oneStepRewrite[times[e_,f_]] /; notIntNotSym[e]&&notIntNotSym[f] := {times[oneStepRewrite[e][[1]],oneStepRewrite[f][[1]]]};
(*
oneStepRewrite[times[x, times[0, pow[x, -1]]]]*)

(* Power - "pow" *)
(* Your code here... *)
oneStepRewrite[pow[e_,0]]:={1};
oneStepRewrite[pow[0,n_Integer?Positive]]:={0};
oneStepRewrite[pow[e_?Positive,f_]]:={Power[e,f]};

(* Derivation - "derivative" *)
(* Your code here... *)
oneStepRewrite[derivative[plus[e_,f_]]]:={plus[derivative[e],derivative[f]]};
oneStepRewrite[derivative[times[e_,f_]]]:={plus[times[e,derivative[f]],times[derivative[e],f]]};
oneStepRewrite[derivative[pow[e_,n_]]]:={times[derivative[e],times[n,pow[e,n-1]]]};



