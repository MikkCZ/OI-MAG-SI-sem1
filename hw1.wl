(* ::Package:: *)

isRewritable[e_]:=(Or[Head[oneStepRewrite[e]]=!=oneStepRewrite,e==={}]);

(* Number *)
oneStepRewrite[derivative[n_Integer]]:={0};

(* Symbol - "x" *)
oneStepRewrite[derivative[x]]:={1};

(* Addition - "plus" *)
(* Your code here... *)
oneStepRewrite[plus[e_?PossibleZeroQ,f_]]:={f};
oneStepRewrite[plus[e_,f_?PossibleZeroQ]]:={e};
oneStepRewrite[plus[e_Integer,f_Integer]]:={Plus[e,f]};

(* Multiplication - "times" *)
(* Your code here... *)
oneStepRewrite[times[e_?PossibleZeroQ,f_]]:={0};
oneStepRewrite[times[e_,f_?PossibleZeroQ]]:={0};
oneStepRewrite[times[e_Integer,f_Integer]]:={Times[e,f]};

(* Power - "pow" *)
(* Your code here... *)
oneStepRewrite[pow[e_,0]]:={1};
oneStepRewrite[pow[0,n_Integer?Positive]]:={0};
oneStepRewrite[pow[e_,f_]]:={Power[e,f]};

(* Derivation - "derivative" *)
(* Your code here... *)
oneStepRewrite[derivative[plus[e_,f_]]]:={plus[derivative[e],derivative[f]]};
oneStepRewrite[derivative[times[e_,f_]]]:={plus[times[e,derivative[f]],times[derivative[e],f]]};
oneStopRewrite[derivative[pow[e_,n_]]]:={times[derivative[e],times[n,pow[e,n-1]]]};
