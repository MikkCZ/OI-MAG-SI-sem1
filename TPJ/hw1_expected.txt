oneStepRewrite[42]
oneStepRewrite[x]
{0}
{1}
{4}
oneStepRewrite[times[2, x]]
oneStepRewrite[times[x, 2]]
oneStepRewrite[times[x, x]]
{0}
{0}
{0}
{0}
{plus[times[2, derivative[2]], times[derivative[2], 2]]}
{plus[times[2, derivative[x]], times[derivative[2], x]]}
{plus[times[x, derivative[2]], times[derivative[x], 2]]}
{plus[times[x, derivative[x]], times[derivative[x], x]]}
{4}
oneStepRewrite[plus[2, x]]
oneStepRewrite[plus[x, 2]]
oneStepRewrite[plus[x, x]]
{2}
{2}
{x}
{x}
{plus[derivative[2], derivative[2]]}
{plus[derivative[2], derivative[x]]}
{plus[derivative[x], derivative[2]]}
{plus[derivative[x], derivative[x]]}
{plus[times[plus[1, x], derivative[x]], times[derivative[plus[1, x]], x]]}
{plus[times[plus[1, x], 1], times[derivative[plus[1, x]], x]], plus[times[plus[1, x], derivative[x]], times[plus[derivative[1], derivative[x]], x]]}
{1}
{1}
{0}
oneStepRewrite[pow[0, -1]]
{1}
oneStepRewrite[pow[plus[x, 42], 2]]
oneStepRewrite[pow[plus[times[x, 2], 42], 2]]
{times[derivative[plus[times[x, 2], 42]], times[2, pow[plus[times[x, 2], 42], 1]]]}
{times[derivative[0], times[0, pow[0, -1]]]}
{times[derivative[1], times[0, pow[1, -1]]]}
{times[derivative[0], times[1, pow[0, 0]]]}
{times[derivative[0], times[-1, pow[0, -2]]]}
{times[derivative[x], times[0, pow[x, -1]]]}
{times[x, 0]}
